src/Compilers/CSharp/Portable/Binder/Binder_TupleOperators.cs

```diff
@@ -0,0 +1,282 @@
+
+            TypeSymbol resultType = GetSpecialType(SpecialType.System_Boolean, diagnostics, node);
+            return new BoundTupleBinaryOperator(node, convertedLeft, convertedRight, kind, operators, resultType);
+        }
+
+        private TupleBinaryOperatorInfo BindTupleBinaryOperatorInfo(BinaryExpressionSyntax node, BinaryOperatorKind kind,
+            BoundExpression left, BoundExpression right, DiagnosticBag diagnostics)
+        {
+            int? leftCardinality = GetTupleCardinality(left);
+            int? rightCardinality = GetTupleCardinality(right);
+
+            if (leftCardinality.HasValue && rightCardinality.HasValue)
+            {
+                TypeSymbol leftType = left.Type;
+                TypeSymbol rightType = right.Type;
+
+                if (leftCardinality.Value != rightCardinality.Value)
+                {
+                    Error(diagnostics, ErrorCode.ERR_TupleSizesMismatchForBinOps, node, leftCardinality, rightCardinality);
+
+                    return new TupleBinaryOperatorInfo.Single(leftType ?? CreateErrorType(), rightType ?? CreateErrorType(),
+                        BinaryOperatorKind.Error, methodSymbolOpt: null, boolConversion: Conversion.NoConversion, boolOperator: default);
+                }
+
+                // typeless tuple literals are not nullable
+                bool leftNullable = leftType?.IsNullableType() == true;
+                bool rightNullable = rightType?.IsNullableType() == true;
+
+                return BindTupleBinaryOperatorNestedInfo(node, kind,
+                    GetTupleArgumentsOrPlaceholders(left), GetTupleArgumentsOrPlaceholders(right),
+                    leftNullable || rightNullable, diagnostics);
+            }
+
+            return BindTupleBinaryOperatorSingleInfo(node, kind, left, right, diagnostics);
+        }
+
+        private TupleBinaryOperatorInfo BindTupleBinaryOperatorSingleInfo(BinaryExpressionSyntax node, BinaryOperatorKind kind,
+            BoundExpression left, BoundExpression right, DiagnosticBag diagnostics)
+        {
+            TypeSymbol leftType = left.Type;
+            TypeSymbol rightType = right.Type;
+
+            if ((object)leftType != null && leftType.IsDynamic() || (object)rightType != null && rightType.IsDynamic())
+            {
+                return BindTupleDynamicBinaryOperatorSingleInfo(node, kind, left, right, diagnostics);
+            }
+
+            LookupResultKind resultKind;
+            BinaryOperatorAnalysisResult best = this.BinaryOperatorOverloadResolution(kind, left, right, node, diagnostics, out resultKind, out var _);
+
+            MethodSymbol resultMethod = null;
+            BinaryOperatorKind resultOperatorKind;
+            TypeSymbol convertedLeftType;
+            TypeSymbol convertedRightType;
+            TypeSymbol returnType;
+            bool hasErrors;
+
+            if (!best.HasValue)
```

Comment:
>if (!best.HasValue) [](start = 12, length = 19)

It feels like there is rather significant amount of code duplication with BindSimpleBinaryOperator method, quite possibly with some deviations. If the logic should be different, I would want that to be stated explicitly in a comment, otherwise I think we should figure out how to factor out and reuse common logic between two methods.

Reply:
I've tried avoiding duplication, but didn't figure out a better solution so far. I'm open to suggestions. 
I'll add a PROTOTYPE marker and we can discuss after I explore folding the binding of tuple operators into `BinaryOperatorOverloadResolution`.

---
In reply to: [170492655](https://github.com/dotnet/roslyn/pull/23356#discussion_r170492655) [](ancestors = 170492655)